MapleServer2/PacketHandlers/Game/ItemSocketSystemHandler.cs

```diff
@@ -30,32 +42,191 @@ public override void Handle(GameSession session, PacketReader packet)
+            List<long> fodderUids = new List<long>();
+            for (int i = 0; i < fodderAmount; i++)
+            {
+                long fodderUid = packet.ReadLong();
+                fodderUids.Add(fodderUid);
+            }
 
             if (!session.Player.Inventory.Items.ContainsKey(itemUid))
             {
+                session.Send(ItemSocketSystemPacket.Notice((int) ItemSocketSystemNotice.ItemIsNotInYourInventory));
                 return;
             }
+            Item equip = session.Player.Inventory.Items[itemUid];
+            int equipUnlockedSlotCount = equip.Stats.GemSockets.Where(x => x.IsUnlocked == true).Count();
 
-            Item gem = session.Player.Inventory.Items[itemUid];
+            foreach (long uid in fodderUids)
+            {
+                if (!session.Player.Inventory.Items.ContainsKey(uid))
+                {
+                    session.Send(ItemSocketSystemPacket.Notice((int) ItemSocketSystemNotice.ItemIsNotInYourInventory));
+                    return;
+                }
+
+                Item fodder = session.Player.Inventory.Items[uid];
+                int fodderUnlockedSlotCount = fodder.Stats.GemSockets.Where(x => x.IsUnlocked == true).Count();
+                if (equipUnlockedSlotCount != fodderUnlockedSlotCount)
+                {
+                    session.Send(ItemSocketSystemPacket.Notice((int) ItemSocketSystemNotice.CannotBeUsedAsMaterial));
+                    return;
+                }
+            }
+
+            // get socket slot to unlock
+            int slot = equip.Stats.GemSockets.FindIndex(0, equip.Stats.GemSockets.Count, x => x.IsUnlocked != true);
+            if (slot < 0)
+            {
+                return;
+            }
+
+            // fragmment cost. hard coded into the client?
+            int crystalFragmentCost = 0;
+            if (slot == 0)
+            {
+                crystalFragmentCost = 400;
+            }
+            else if (slot == 1 || slot == 2)
+            {
+                crystalFragmentCost = 600;
+            }
+
+            int crystalFragmentsTotalAmount = 0;
+            List<KeyValuePair<long, Item>> crystalFragments = session.Player.Inventory.Items.Where(x => x.Value.Tag == "CrystalPiece").ToList();
+            crystalFragments.ForEach(x => crystalFragmentsTotalAmount += x.Value.Amount);
+
+            if (crystalFragmentsTotalAmount < crystalFragmentCost)
+            {
+                return;
+            }
+
+            foreach (KeyValuePair<long, Item> item in crystalFragments)
+            {
+                if (item.Value.Amount >= crystalFragmentCost)
+                {
+                    InventoryController.Consume(session, item.Key, crystalFragmentCost);
+                    break;
+                }
+                else
+                {
+                    crystalFragmentCost -= item.Value.Amount;
+                    InventoryController.Consume(session, item.Key, item.Value.Amount);
+                }
```

Comment:
Isn't line 141 useless? and can't you just use math.max?

Reply:
I don't think line 141 is useless. It's needed to update the value of how many crystal fragments are left in the event that the item amount < cost.

Reply:
`crystalFragmentCost` isn't used after the value is changed on line 141.  so it's useless 🤔

Reply:
it's a loop tho so it is used. or am I misunderstanding?

Reply:
welp im blind,